VogtInterface/ProductsBase.cs

```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace VogtInterface
+{
+    abstract class ProductsBase
+    {
+        public abstract void PrintProductDetails(List<IProductsModel> products);
```

Comment:
There are two possible solutions here. It depends on whether each specific type of product will **definitely** have its own type of info to print. If so, then I would force EACH specific type of product to implement its own details.

# Solution 1

This solution will an interface to expose the PrintProductDetails function to other classes. However, it will use an abstract function to force each specific model to implement the specific details.

## Function 1
The purpose of this function is to print the **specific** type of product information. Therefore, this is abstract and each concrete class will need to implement it.
```c3
protected abstract void PrintProductSpecificDetails();
```

## Function 2
```c#
public void PrintProductDetails() {
    Console.WriteLine($"Product name: {product.Name}");
    Console.WriteLine($"Product price: {product.Price:C2}");
    PrintProductSpecificDetails();
}
```

## Function 3

This is an example of a specific model class then implementing the abstract function.

```c#
public override void PrintProductSpecificDetails() {
    Console.WriteLine($"Meat type: {Type}");
}
```

## Summary

The specific model classes are forced to implement the abstract function. The base class, which should be in your interface (void PrintProductDetails) i.e. `Function 2` is used by external classes and is the only function exposed publically.

# Solution 2

If only SOME model classes need to print custom detail, then I wouldn't force them to implement the abstract method. Instead, I'd only implement a single **virtual** method in the base class:

```c#
public virtual void PrintProductDetails() {
    Console.WriteLine($"Product name: {product.Name}");
    Console.WriteLine($"Product price: {product.Price:C2}");
}
```

Then, you can **override** it only in the models where you need it. For example, the MeatModel could override it by doing:
```c#
public override void PrintProductDetails() {
    base.PrintProductDetails();
    Console.WriteLine($"Meat type: {Type}");
}
```

Reply:
Btw, once you implement either of the above solutions.. you should be able to do this in `Program.cs`:

```c#
foreach (var product in products) {
    product.PrintProductDetails();
}
```

In both solutions -- the outcome should be the same. The difference is the design.